core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java

```diff
@@ -0,0 +1,196 @@
+    private final UniformPartitionRebalancer partitionRebalancer;
+    private final PageReleasedListener onPageReleased;
+
+    private final IntArrayList[] partitionAssignments;
+    private final IntArrayList[] writerAssignments;
+    private final int[] partitionWriterIndex;
+
+    // Use Long2LongMap instead of Map<WriterPartitionId, Long> which helps to save memory in the worst case scenario.
+    // Here first 32 bit of long key contains writerId whereas last 32 bit contains partitionId.
+    @GuardedBy("this")
+    private final Long2LongMap partitionsRowCount = new Long2LongOpenHashMap();
+
+    public ScaleWriterPartitioningExchanger(
+            List<Consumer<PageReference>> buffers,
+            LocalExchangeMemoryManager memoryManager,
+            long maxBufferedBytes,
+            Function<Page, Page> partitionedPagePreparer,
+            PartitionFunction partitionFunction,
+            int partitionCount,
+            UniformPartitionRebalancer partitionRebalancer)
+    {
+        this.buffers = requireNonNull(buffers, "buffers is null");
+        this.memoryManager = requireNonNull(memoryManager, "memoryManager is null");
+        this.maxBufferedBytes = maxBufferedBytes;
+        this.partitionedPagePreparer = requireNonNull(partitionedPagePreparer, "partitionedPagePreparer is null");
+        this.partitionFunction = requireNonNull(partitionFunction, "partitionFunction is null");
+        this.partitionRebalancer = requireNonNull(partitionRebalancer, "partitionRebalancer is null");
+        this.onPageReleased = forLocalExchangeMemoryManager(memoryManager);
+
+        // Initialize writerAssignments with the buffer size
+        writerAssignments = new IntArrayList[buffers.size()];
+        for (int i = 0; i < writerAssignments.length; i++) {
+            writerAssignments[i] = new IntArrayList();
+        }
+
+        // Initialize partitionAssignments and partitionWriterIndex with the partition limit
+        partitionWriterIndex = new int[partitionCount];
+        partitionAssignments = new IntArrayList[partitionCount];
+        for (int i = 0; i < partitionAssignments.length; i++) {
+            partitionAssignments[i] = new IntArrayList();
+        }
+    }
+
+    @Override
+    public void accept(Page page)
+    {
+        Page partitionPage = partitionedPagePreparer.apply(page);
+
+        // Assign each row to a partition which limits to the partitionCount. If there are more physical partitions than
+        // this artificial partition limit, then it is possible that multiple partitions will get assigned the same
+        // bucket id. Thus, multiple partitions will be scaled together since we track partition physicalWrittenBytes
+        // using the artificial limit (partitionCount). The assignments lists are all expected to cleared by the
+        // previous iterations.
+        for (int position = 0; position < partitionPage.getPositionCount(); position++) {
```

Comment:
you can change `partitionAssignments` as `int[] partitionRowCounts`.
You can also have `int partitionWriterIds[]`

Then you could do:
```
for (int position = 0; position < partitionPage.getPositionCount(); position++) {
  int partitionId = partitionFunction.getPartition(partitionPage, position);
  partitionRowCounts[partition] = partitionRowCount[partitionId] + 1;
  int writerId = partitionWriterIds[partitionId];
  if (writerId == -1) {
     writerId = getNextWriterId(partitionId);
     partitionWriterIds[partitionId] = writerId;
  }
  writerAssignments[writerId].add(positions[i]);
}
fill(partitionWriterIds, -1)
```

Reply:
Generally, I think you could even collect `Long2IntMap pagePartitionRowCounts` directly here.
Otherwise you would still have to iterate though all `partitionRowCounts` (4096 positions) even if your pages are small (e.g. 200 positions)

Reply:
Another option is to do:
```
IntArrayList usedPartitions; // pre-allocated
for (int position = 0; position < partitionPage.getPositionCount(); position++) {
  int partitionId = partitionFunction.getPartition(partitionPage, position);
  int writerId = partitionWriterIds[partitionId] * epoch;
  if (writerId < 0) {
     writerId = getNextWriterId(partitionId);
     partitionWriterIds[partitionId] = writerId * epoch;
     partitionRowCounts[partitionId] = 1;
     usedPartitions.add(partitionId);
  } else {
     partitionRowCounts[partitionId] = partitionRowCount[partitionId] + 1;
  }
  writerAssignments[writerId].add(positions[i]);
}
epoch *= -1;
```
Then you don't even need to do `fill(partitionWriterIds, -1)`